server/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -366,8 +420,20 @@ public AllocationStatus getLastAllocationStatus() {
      *
      * @return calculated delay in nanoseconds
      */
-    public long getRemainingDelay(final long nanoTimeNow, final Settings indexSettings) {
-        long delayTimeoutNanos = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexSettings).nanos();
+    public long getRemainingDelay(
+        final long nanoTimeNow,
+        final Settings indexSettings,
+        final Map<String, SingleNodeShutdownMetadata> nodesShutdownMap
+    ) {
+        final long indexLevelDelay = INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.get(indexSettings).nanos();
+        long delayTimeoutNanos = Optional.ofNullable(lastAllocatedNodeId)
+            .filter(nodeId -> lastAllocatedNodeIsRestarting) // If the node wasn't restarting when this became unassigned, use default delay
+            .map(nodesShutdownMap::get)
+            .filter(shutdownMetadata -> SingleNodeShutdownMetadata.Type.RESTART.equals(shutdownMetadata.getType()))
+            .map(SingleNodeShutdownMetadata::getShardReallocationDelay)
+            .map(TimeValue::nanos)
```

Comment:
nit: The auto-boxing here annoys me slightly, I would prefer to terminate the optional map'ing here and work on the longs. We need Optional.mapToLong :slightly_smiling_face:. I can live with it as is though, escape analysis might catch it and otherwise it is unlikely to be a big deal (I think).

Reply:
Given that it 1) gets immediately auto-unboxed back to a primitive `long` and 2) is not possible to become null before it gets unboxed, I'm not sure what we lose here other than an unnecessary object allocation or two? (given that it'll be unboxed for the `Math.max` call and reboxed into the optional afterwards)

And it seems to me the increase in readability outweighs the slight inefficiency.